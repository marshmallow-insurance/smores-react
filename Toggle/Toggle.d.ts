import { FC } from 'react';
import { MarginProps } from '../utils/space';
type Props = {
    /** unique ID */
    id?: string;
    /** checked flag  */
    checked: boolean;
    /** onToggle listener  */
    onToggle: () => void;
} & MarginProps;
export declare const Toggle: FC<Props>;
export {};
