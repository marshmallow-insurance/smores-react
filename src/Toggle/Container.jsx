import React, {useState} from 'react';
import {Toggle} from './Toggle';

export const Container = () => {
  const [active, setActive] = useState(false);

  return <Toggle checked={active} onToggle={() => setActive(!active)} />;
};
