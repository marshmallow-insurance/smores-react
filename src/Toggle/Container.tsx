import React, { useState } from 'react'
import { Toggle } from './Toggle'

export const Container = () => {
  const [active, setActive] = useState(false)

  return (
    <Toggle
      id="toggleId"
      checked={active}
      onToggle={() => setActive(!active)}
    />
  )
}
