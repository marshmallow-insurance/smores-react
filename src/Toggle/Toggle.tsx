import React, { FC } from 'react'
import styled from 'styled-components'

import { focusOutline } from '../utils/focusOutline'
import { theme } from '../theme'

type Props = {
  /** unique ID */
  id?: string
  /** checked flag  */
  checked: boolean
  /** onToggle listener  */
  onToggle: () => void
}

export const Toggle: FC<Props> = ({ checked, onToggle }) => {
  return (
    <Switch>
      <Checkbox type="checkbox" checked={checked} onChange={onToggle} />
      <Slider />
    </Switch>
  )
}

const Switch = styled.label`
  position: relative;
  display: inline-block;
  width: 50px;
  height: 32px;

  input {
    opacity: 0;
    width: 0;
    height: 0;
  }
`

const Slider = styled.span`
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: ${theme.colors.bgPrimary};
  border: 1px solid;
  border-color: ${theme.colors.bgPrimary};
  border-radius: 28px;
  transition: 0.2s background-color border-color;
  outline: none;

  &:before {
    position: absolute;
    content: '';
    height: 24px;
    width: 24px;
    left: 3px;
    bottom: 3px;
    background-color: ${theme.colors.white};
    transition: 0.2s transform;
    border-radius: 50%;
  }
`

const Checkbox = styled.input`
  ${focusOutline({ selector: `&:focus-visible + ${Slider}` })}

  &:checked + ${Slider} {
    background-color: ${theme.colors.success};
    border: 1px solid;
    border-color: ${theme.colors.success};
  }

  &:checked + ${Slider}:before {
    transform: translateX(18px);
  }
`
