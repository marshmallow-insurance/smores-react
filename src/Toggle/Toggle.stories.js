import React from 'react'
import { storiesOf } from '@storybook/react'

import { Container } from './Container.jsx'

storiesOf('Toggle', module).addWithJSX('default', () => <Container />)
