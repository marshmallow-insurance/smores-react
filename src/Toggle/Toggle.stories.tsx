import React from 'react'
import { Toggle } from './Toggle'
import { Container } from './Container'

export default {
  title: 'Toggle',
  component: Toggle,
}

const Template = () => <Container />

export const Default = Template.bind({})
