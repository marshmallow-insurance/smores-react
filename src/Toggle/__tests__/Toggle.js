import React from 'react';
import {render} from '@testing-library/react';
import 'jest-styled-components';

import {Toggle} from '../Toggle';

test('renders', () => {
  const {container} = render(
    <Toggle id="toggle" checked={false} onToggle={() => {}} />,
  );
  expect(container.firstChild).toMatchSnapshot();
});
