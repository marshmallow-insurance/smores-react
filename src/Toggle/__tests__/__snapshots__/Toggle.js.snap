// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  width: 56px;
  height: 32px;
}

.c0 input {
  opacity: 0;
  width: 0;
  height: 0;
}

.c3 {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #DAD2C4;
  border: none;
  border-radius: 28px;
  -webkit-transition: 0.2s background-color;
  transition: 0.2s background-color;
  outline: none;
}

.c3:hover {
  background-color: #A1957C;
}

.c3:before {
  position: absolute;
  content: '';
  height: 24px;
  width: 24px;
  left: 4px;
  bottom: 4px;
  background-color: #F8F2EA;
  -webkit-transition: 0.2s -webkit-transform;
  -webkit-transition: 0.2s transform;
  transition: 0.2s transform;
  border-radius: 50%;
}

.c1 {
  outline: 0;
}

.c1:focus-visible + .c2 {
  outline: 2px solid #292924;
  outline-offset: 2px;
}

.c1:checked + .c2 {
  background-color: #292924;
  border: none;
}

.c1:checked + .c2:before {
  -webkit-transform: translateX(24px);
  -ms-transform: translateX(24px);
  transform: translateX(24px);
}

<label
  class="c0"
  id="toggle"
>
  <input
    class="c1"
    type="checkbox"
  />
  <span
    class="c2 c3"
  />
</label>
`;
